pkg/util/scheduler/queue_test.go

```diff
@@ -0,0 +1,635 @@
+	})
+
+	t.Run("TenantQueueFullError", func(t *testing.T) {
+		t.Parallel()
+
+		ctx := context.Background()
+		q := NewQueue(QueueOptionsWithDefaults(&QueueOptions{MaxSizePerTenant: 2}))
+		require.NoError(t, q.StartAsync(context.Background()), "Queue should start")
+		require.NoError(t, q.AwaitRunning(context.Background()), "Queue should be running")
+
+		tenantID := "tenant-limited"
+
+		// Enqueue up to the limit
+		err := q.Enqueue(ctx, tenantID, func() {})
+		require.NoError(t, err)
+		err = q.Enqueue(ctx, tenantID, func() {})
+		require.NoError(t, err)
+
+		require.Equal(t, 2, q.Len())
+		require.Equal(t, 1, q.ActiveTenantsLen())
+
+		// Enqueue one more, expect error
+		err = q.Enqueue(ctx, tenantID, func() {})
+		require.ErrorIs(t, err, ErrTenantQueueFull, "Expected ErrTenantQueueFull")
+
+		// Len should still be 2
+		require.Equal(t, 2, q.Len())
+
+		// Dequeue one item
+		dequeueCtx, cancel := context.WithTimeout(ctx, 50*time.Millisecond)
+		_, ok, err := q.Dequeue(dequeueCtx)
+		cancel()
+		require.NoError(t, err)
+		require.True(t, ok)
+		require.Equal(t, 1, q.Len())
+
+		// Now enqueue should succeed again
+		err = q.Enqueue(ctx, tenantID, func() {})
+		require.NoError(t, err, "Enqueue should succeed after dequeueing one item")
+		require.Equal(t, 2, q.Len(), "Length should be back to 2")
+	})
+
+	t.Run("DequeueContextCancellation", func(t *testing.T) {
+		t.Parallel()
+
+		ctx := context.Background()
+		q := NewQueue(QueueOptionsWithDefaults(nil))
+		require.NoError(t, q.StartAsync(ctx), "Queue should start")
+		require.NoError(t, q.AwaitRunning(ctx), "Queue should be running")
+
+		// Create an already canceled context instead of using timeout
+		cancelCtx, cancel := context.WithCancel(context.Background())
+		cancel() // Cancel immediately
+
+		// This should return with context.Canceled error immediately
+		runnable, ok, err := q.Dequeue(cancelCtx)
+
+		// Verify the result - should be context.Canceled, not DeadlineExceeded
+		require.Nil(t, runnable, "Runnable should be nil on context cancellation")
+		require.False(t, ok, "ok should be false on context cancellation")
+		require.ErrorIs(t, err, context.Canceled, "Expected context canceled error")
+	})
+
+	t.Run("DequeueWithTimeout", func(t *testing.T) {
+		t.Parallel()
+
+		ctx := context.Background()
+		q := NewQueue(QueueOptionsWithDefaults(nil))
+		require.NoError(t, q.StartAsync(ctx), "Queue should start")
+		require.NoError(t, q.AwaitRunning(ctx), "Queue should be running")
+
+		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		defer cancel()
+
+		// This should timeout since nothing is in the queue
+		runnable, ok, err := q.Dequeue(ctx)
+		require.Nil(t, runnable, "Runnable should be nil on timeout")
+		require.False(t, ok, "ok should be false on timeout")
+		require.ErrorIs(t, err, context.DeadlineExceeded, "Expected context deadline exceeded error")
+	})
+
+	t.Run("EnqueueAfterStopped", func(t *testing.T) {
```

Comment:
What about enqueue before start? Perhaps `Enqueue` should check if service is in `Running` state?